include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle.hpp

```diff
@@ -397,162 +130,80 @@ struct DeviceGemm_Xdl_CShuffle : public DeviceGemm<ALayout,
             {
                 throw std::runtime_error("wrong! GridwiseGemm has invalid setting");
             }
 
-            const index_t grid_size =
-                arg.block_2_ctile_map_.CalculateGridSize(arg.c_grid_desc_m_n_);
-            const auto K =
-                arg.a_grid_desc_ak0_m_ak1_.GetLength(I0) * arg.a_grid_desc_ak0_m_ak1_.GetLength(I2);
+            index_t gdx, gdy, gdz;
+            std::tie(gdx, gdy, gdz) = GridwiseGemm::CalculateGridSize(karg.M, karg.N);
+
+            const auto K = GridwiseGemm::CalculateAK0(karg.K) * AK1;
 
             float ave_time = 0;
 
             if(GridwiseGemm::CalculateHasMainKBlockLoop(K))
             {
-                const auto kernel = kernel_gemm_xdl_cshuffle_v1<
-                    GridwiseGemm,
-                    ADataType, // TODO: distiguish A/B datatype
-                    CDataType,
-                    AElementwiseOperation,
-                    BElementwiseOperation,
-                    CElementwiseOperation,
-                    DeviceOp::AGridDesc_AK0_M_AK1,
-                    DeviceOp::BGridDesc_BK0_N_BK1,
-                    typename GridwiseGemm::CGridDescriptor_MBlock_MPerBlock_NBlock_NPerBlock,
-                    typename GridwiseGemm::DefaultBlock2CTileMap,
-                    true>;
-
-                ave_time =
-                    launch_and_time_kernel(stream_config,
-                                           kernel,
-                                           dim3(grid_size),
-                                           dim3(BlockSize),
-                                           0,
-                                           arg.p_a_grid_,
-                                           arg.p_b_grid_,
-                                           arg.p_c_grid_,
-                                           arg.a_element_op_,
-                                           arg.b_element_op_,
-                                           arg.c_element_op_,
-                                           arg.a_grid_desc_ak0_m_ak1_,
-                                           arg.b_grid_desc_bk0_n_bk1_,
-                                           arg.c_grid_desc_mblock_mperblock_nblock_nperblock_,
-                                           arg.block_2_ctile_map_);
+                const auto kernel = kernel_gemm_xdl_cshuffle_v1_simplified<GridwiseGemm, true>;
+
+                ave_time = launch_and_time_kernel(stream_config,
+                                                  kernel,
+                                                  dim3(gdx, gdy, gdz),
+                                                  dim3(BlockSize),
+                                                  0,
+                                                  karg.p_a_grid,
+                                                  karg.p_b_grid,
+                                                  karg.p_c_grid,
+                                                  karg);
```

Comment:
Here you're duplicating tensor pointers as kernel arguments, since they're already inside `karg` argument. Maybe there's no need to store them in DeviceOp::Argument level ?

Reply:
But we need to provide an interface for users to send those pointers, I think it's what `DeviceOp::MakeArgument()` does.

To avoid re-assign pointer before launching kernel in DeviceCGemm_4Gemm_Xdl_CShuffle<>, I didn't wrap the pointers into `GridwiseGemm::Argument` either.

Is there a better place to put the pointers to?

Reply:
The user either way has to provide `DeviceOp::Argument` as an argument to the kernel function. This argument object should have already properly set pointers since this is what it's done during it's constructions when you call `MakeArgument`. The proper place for the pointers is `Argument` object, thus there's no need to redundantly and separately pass them to the kernel function.

Reply:
One more thing came to my mind. Right now those pointers are taken as `__restrict__` pointers in kernel function, which is done for compiler optimization reasons. I wonder if we would pass them within `Argument` object if that would hinder the compiler job. But I guess there shouldn't be any problem to have `__restrict__` pointer members in `Argument` object.

Reply:
I think I should create another kernel for `DeviceCGemm_4Gemm_Xdl_CShuffle<>` to send the pointer arguments separately